Allow to override the return type of FFI functions

ergl opened the pull request:
When this is merged, the commit message shoud be the following:

---

Allow overriding the return type of FFI functions

On the sync call for March 1st, we decided to re-add the ability to specify a
return type for FFI function calls. This was previously removed by RFC 68 on the
grounds that, with FFI declarations being mandatory, specifying a return type
would be redundant.

Before RFC 68 was implemented, the return type at the call site was not allowed
to differ from the return type of the declaration. We have lifted this
limitation, which should help users write FFI bindings for C functions that
might return different types depending on how they are called.

The return type of an FFI function without an explicit return type will default
to the return type of its declaration.

ponylang-main commented:
<!-- release-note-reminder-bot -->
Hi @ergl,

The changelog - added label was added to this pull request; all PRs with a changelog label need to have release notes included as part of the PR. If you haven't added release notes already, please do.

Release notes are added by creating a uniquely named file in the `.release-notes` directory. We suggest you call the file `4060.md` to match the number of this pull request.

The basic format of the release notes (using markdown) should be:

```
## Title

End user description of changes, why it's important,
problems it solves etc.

If a breaking change, make sure to include 1 or more
examples what code would look like prior to this change
and how to update it to work after this change.
```

Thanks.

SeanTAllen commented:
Sweet

SeanTAllen commented:
Are there any tutorial updates that need to go along with this?

ergl commented:
@SeanTAllen We should probably add a note about this in the tutorial, although what's currently there is still valid.

SeanTAllen commented:
@ergl can you handle the "sounds like it is small" tutorial PR?

ergl commented:
@SeanTAllen Yeah, no problem.

SeanTAllen commented on .release-notes/4060.md:
```diff
@@ -0,0 +1,7 @@
+## Allow overriding the return type of FFI functions
+
+On the sync call for March 1st, we decided to re-add the ability to specify a return type for FFI function calls. This was previously removed by RFC 68 on the grounds that, with FFI declarations being mandatory, specifying a return type would be redundant.
```
Is there a reason for calling out the sync call? that seems like extra, unneeded info. if it is because people could go to listen to  get more context, i think we should try to provide full context in the release notes.

I'd like to change these release notes to show something from red or someone else that was brought up in zulip that was a completely valid use case and wasn't possible due to the change.

ergl replied:
Sure, my bad, I blindly copy-pasted the commit message into the release notes. Let me rework it a bit.

ergl commented:
@SeanTAllen I've reworked the release notes, and added some examples of what this change makes possible. Let me know if you think it's too long/verbose, and I can try to trim it a bit.

SeanTAllen reviewed (approved):
I like the release notes.

ergl commented on .release-notes/4060.md:
````diff
@@ -0,0 +1,78 @@
+  var y: U64 = 0
+```
+
+The problem with these declarations is that the `list_push` and `list_pop` function can only operate with lists of elements of type `Point2D`. If we wanted to use these functions with different types, we would modify `list_push` to take an element of type `Pointer[None]`, which represents the `void*` type. Until now, however, there was no way of making `list_pop` truly generic:
````
```suggestion
The problem with these declarations is that the `list_push` and `list_pop` functions can only operate with lists of elements of type `Point2D`. If we wanted to use these functions with different types, we would modify `list_push` to take an element of type `Pointer[None]`, which represents the `void*` type. Until now, however, there was no way of making `list_pop` truly generic:
```

SeanTAllen commented:
Before merging, I'd like @jemc to sign off on this to verify that it matches his expectation of what we agreed to do for re-enabling this functionality.

jemc commented on test/libponyc-run/ffi-different-returns/main.pony:
```diff
@@ -0,0 +1,12 @@
+use "lib:ffi-different-returns-additional"
+
+use @ffi_function[I32]()
+use @pony_exitcode[None](code: I32)
+
+actor Main
+  new create(env: Env) =>
+    let res: (I32 | U8) = @ffi_function[U8]()
```
This actually needs to be disallowed - the differing return type needs to be ABI-compatible with all other invocations of the same function. Specifically, I think we need to enforce that all of them are the same ABI size.

This is easy for pointers and classes, because they will all be pointer-sized.

But in this case you have a 1-byte and a 4-byte variant, which are not compatible.

SeanTAllen replied:
So that was a bug we had previously with return types for FFI?

SeanTAllen replied:
definitely something we need to make sure we have tests for then. and we should update the release notes and tutorial to explain the limitation.

thanks for catching that. it makes sense as a limitation but i wasn't thinking of it when i reviewed.

ergl replied:
After discussing this with @jemc, we agreed that this check is "good enough" for our purposes, as it will catch multiple calls to the same FFI function with incompatible return types.

It's not feasible to perform the ABI size check against the return type as specified in the FFI declaration without rearchitecting how we perform the code generation for FFI functions in a significant way.

The way that the implementation works, the compiler will treat the first FFI call to a function as "authoritative", and all future calls to this function will be checked for ABI compatibility against it. This didn't cause a problem before because type checking was making sure that the return type for calls matched the return type of the declaration.

ergl commented:
I'll add a note to the release notes about the ABI compatiblity. Other than that, I think this is good to merge.
